compiler/optimizations/remoteValueForwarding.cpp

```diff
@@ -208,10 +209,107 @@ static void updateTaskFunctions(Map<Symbol*, Vec<SymExpr*>*>& defMap,
       for_formals(arg, fn) {
         if (canForwardValue(defMap, useMap, syncSet, fn, arg)) {
           updateTaskArg(useMap, fn, arg);
+          AggregateType* argType = toAggregateType(arg->type);
+          if (argType && isRecord(argType)) {
+            for_fields(field, argType) {
+              if (isRecordWrappedType(field->getValType()) && field->isRef()) {
+                if (fieldsToRVF.find(arg) == fieldsToRVF.end()) {
+                  fieldsToRVF[arg] = std::vector<Symbol*>();
+                }
+                fieldsToRVF[arg].push_back(field);
+              }
+            }
+          }
+        }
+      }
+
+      if (fieldsToRVF.size() > 0) {
+        SET_LINENO(fn);
+        AggregateType* wrapType = new AggregateType(AGGREGATE_RECORD);
+        TypeSymbol* wrapSym = new TypeSymbol("RVF_wrapType", wrapType);
+        fn->getModule()->block->insertAtTail(new DefExpr(wrapSym));
+        // Do we need to create a ref type at this point?
+        //getOrMakeRefTypeDuringCodegen(wrapper);
+        ArgSymbol* wrapArg = new ArgSymbol(INTENT_CONST_IN, "RVF_wrapper", wrapType);
+        fn->insertFormalAtTail(wrapArg);
+
+        // Map from fields in the original record to fields in the RVF wrapper
+        std::map<Symbol*, Symbol*> fieldToField;
+
+        //
+        // Create the fields in the wrapper record and unpack them within the
+        // on-statement's body.
+        //
+
+        // Use a counter to make the new fields have unique names
+        int counter = 0;
+
+        // Use this call as an anchor point for insertBefores
+        // BHARSH: I did this because I find the insertBefore style easier to
+        // follow
+        CallExpr* anchor = new CallExpr(PRIM_RT_ERROR, new_CStringSymbol("Do not leave this call in the tree"));
+        fn->body->insertAtHead(anchor);
+        for (argIter it = fieldsToRVF.begin(); it != fieldsToRVF.end(); ++it) {
+          ArgSymbol* origArg = toArgSymbol(it->first);
+          for_vector(Symbol, oldField, it->second) {
+            VarSymbol* rvfField = new VarSymbol(astr("RVF_", istr(counter), "_", oldField->cname), oldField->getValType());
+            INT_ASSERT(oldField->isRef() && !rvfField->isRef());
+            wrapType->fields.insertAtTail(new DefExpr(rvfField));
+            fieldToField[oldField] = rvfField;
+            fieldToField[rvfField] = oldField;
+
+            // Get the RVF'd field from the wrapper, and load it into the
+            // original reference field.
+
+            VarSymbol* rvfTemp = newTemp(rvfField->type);
+            anchor->insertBefore(new DefExpr(rvfTemp));
+            anchor->insertBefore(new CallExpr(PRIM_MOVE, rvfTemp, new CallExpr(PRIM_GET_MEMBER_VALUE, wrapArg, rvfField)));
+
+            VarSymbol* refTemp = newTemp(oldField->qualType());
+            anchor->insertBefore(new DefExpr(refTemp));
+            anchor->insertBefore(new CallExpr(PRIM_MOVE, refTemp, new CallExpr(PRIM_SET_REFERENCE, rvfTemp)));
```

Comment:
Also, I wouldn't expect PRIM_SET_REFERENCE to work if refTemp has type _ref(something). I'd expect it to only apply to qual=ref. So perhaps we should consider removing the ref-type part if it is present ?